src/graph_visualization.cpp

```diff
@@ -0,0 +1,463 @@
+#include "graph_visualization.h"
+
+void initIdAndSucc(std::vector<std::vector<int>> &input, std::vector<TreeNode> &nodes);
+void predReinit(std::vector<TreeNode> &nodes);
+void degReinit(std::vector<TreeNode> &nodes);
+
+void algorithmDFS(int &exit, std::vector<TreeNode> &nodes, std::vector<bool> &usedId, TreeNode::nodeId id, TreeNode::nodeId prevId);
+void cycleExistsDFS(std::vector<TreeNode> &nodes);
+
+void heapify(std::vector<TreeNode> &nodes, int n, int i);
+void idHeapSort(std::vector<TreeNode> &nodes);
+TreeNode getTreeNodeById(std::vector<TreeNode> &nodes, TreeNode::nodeId id);
+void removeIncidentArcs(std::vector<TreeNode> &nodes, TreeNode::nodeId id);
+void greedyFAS(std::vector<TreeNode> &nodes, std::vector<std::pair<TreeNode::nodeId, TreeNode::nodeId>> &deletedEdges);
+
+void algorithmASAP(std::vector<TreeNode> &nodes);
+
+void addDummyNode(std::vector<TreeNode> &nodes, TreeNode up, TreeNode down, int order);
+void addDummyNodes(std::vector<TreeNode> &nodes, std::vector<std::pair<TreeNode::nodeId, TreeNode::nodeId>> &deletedEdges);
+
+int main() {
+  std::vector<std::vector<int>> input = {};
+  inputGraph(input);
+
+  std::vector<TreeNode> nodes = {};
+  transformationGraph(input, nodes);
+
+  printInfo(nodes);
+
+  return 0;
+}
+
+void inputGraph(std::vector<std::vector<int>> &input) {
+  for (int i = 0; 1; i++) {
+    int inputNum = -3;
+    std::vector<int> in = {};
+    input.push_back(in);
+    while (1) {
+      scanf("%i", &inputNum);
```

Comment:
Not checking return values of input routines is plainly wrong, do at least
```suggestion
      int rc = scanf("%i", &inputNum);
      assert(rc == 1);
```